### PR TITLE
feat(labware-library): adapter and module support to labware creator

### PR DESCRIPTION
**Pull request by jerader**
closes AUTH-447

# Overview

Adds adapter and module support to custom labware creator in labware library. Now, you can add stacking offsets if you labware fits certain criteria.

# Test Plan

Here is the sandbox: http://sandbox.labware.opentrons.com/ll_add-adapter-field/ - you have to navigate to `Labware Creator` in the Labware Guide on the top left of the screen.

There are a handful of things to test:

- [ ] create a tiprack definition, scroll down to the bottom of the form until you find the `Stacking Offset` section. Only 1 adapter should be visible which is `Opentrons Flex 96 Tip Rack Adapter`. Click on the checkbox and see that 0 is automatically added to the field. Delete it and see that the error banner pops up. Add a number. the error banner should go away
- [ ] create a new definition selecting "wellPlate". Select "v bottom" under `Well Bottom & Depth`. Scroll to `Stacking Offset` and see that 2 adapters are visible and 1 module is visible (the magnetic block v1). Now, scroll up and add a 16mm height, 8 rows that are evenly spaced and 12 columns that are evenly spaced. Scroll to `Stacking Offset` again and see that the Thermocycler GEN2 is now visible.
- [ ]  create a new definition and this time select "reservoir". Some fields will still be populated, that is fine. Scroll to the `Well Bottom & Depth` and select `flat bottom`. Scroll to `Stacking Offset` and see that the `Opentrons 96 Flat Bottom Heater-Shaker Adapter` adapter is visible as well as the magnetic block. Finally, change the `circular` shape to `rectangular` shape and see that 2 adapters are now visible. 

Finally, make a complete definition with the following specs that match this reservoir: https://labware.opentrons.com/biorad_96_wellplate_200ul_pcr
Add 1 adapter and 1 module and set the offsets to 100, export it and you should see something like this for the stacking offset keys:

`  "stackingOffsetWithLabware": {
        "opentrons_96_well_aluminum_block": {
            "x": 0,
            "y": 0,
            "z": -65.84
        }
    },
    "stackingOffsetWithModule": {
        "thermocyclerModuleV2": {
            "x": 0,
            "y": 0,
            "z": -84
        }
    }
`

Import it back and the correct adapters/modules selected should populate with 100 for each field

# Changelog

- create a 2 new fields to formik `compatibleAdapters` and `compatibleModules` that are records of strings and the number that the user inputs
- create a new component that allows users to select the new fields called `StackingOffsets`
- add the stackingoffset keys to `LabwareDefinition2` as optional keys
- add test coverage to `StackingOffsets`
- add conversions from formik form to labware and labware back to formik form (for inputting it back to LC)

# Review requests

see test plan

# Risk assessment

low, LL release happening in August

### CODE REVIEW

**Comment by jerader** [comment]
TODO:

- [x] test coverage
- [x] fix the check error in `labware-library/src/labware-creator/utils/makeAutofillOnChange.ts`
- [ ] add the image that design is still working on

**Comment by jerader** [comment]
> * [ ]  add the image that design is still working on

^ NOTE: not sure when design will have the image ready so I might address that in a follow up PR, depending on when this PR gets approved.

File: labware-library/src/labware-creator/__tests__/StackingOffsets.test.tsx [hunk 1]
```diff
@@ -0,0 +1,103 @@
+import * as React from 'react'
+import { describe, it, expect, vi, beforeEach } from 'vitest'
+import '@testing-library/jest-dom/vitest'
+import { useFormikContext } from 'formik'
+import {
+  getAllDefinitions,
+  fixtureTiprackAdapter,
+  fixture96Plate,
+} from '@opentrons/shared-data'
+import { fireEvent, screen, render } from '@testing-library/react'
```
Comment (koji) [inline labware-library/src/labware-creator/__tests__/StackingOffsets.test.tsx]: this line can be before shared-data

File: labware-library/src/labware-creator/__tests__/StackingOffsets.test.tsx [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import { StackingOffsets } from '../components/sections/StackingOffsets'
+import type * as SharedData from '@opentrons/shared-data'
+import type * as Formik from 'formik'
```
Comment (koji) [inline labware-library/src/labware-creator/__tests__/StackingOffsets.test.tsx]:
```suggestion
import type * as Formik from 'formik'
import type * as SharedData from '@opentrons/shared-data'
```

File: labware-library/src/labware-creator/__tests__/labwareDefToFields.test.ts [hunk 3]
```diff
@@ -6,13 +6,17 @@ import {
...
 } from '@opentrons/shared-data/labware/fixtures/2'
+import type { LabwareDefinition2 } from '@opentrons/shared-data'
 
 vi.mock('../../definitions')
 
 describe('labwareDefToFields', () => {
   it('fixture_96_plate', () => {
     const def = fixture_96_plate
-    const result = labwareDefToFields(def as any)
+    const result = labwareDefToFields(
```
Comment (koji) [inline labware-library/src/labware-creator/__tests__/labwareDefToFields.test.ts:16]: 👍

File: labware-library/src/labware-creator/components/sections/StackingOffsets.tsx [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+import styles from '../../styles.module.css'
+
+const ADAPTER_BLOCK_LIST = ['opentrons_96_deep_well_adapter']
+const MODULE_MODELS_WITH_NO_ADAPTERS: ModuleModel[] = [
+  MAGNETIC_BLOCK_V1,
+  THERMOCYCLER_MODULE_V2,
+]
+
+export const StackingOffsets = (): JSX.Element | null => {
```
Comment (koji) [inline labware-library/src/labware-creator/components/sections/StackingOffsets.tsx]:
```suggestion
export function StackingOffsets(): JSX.Element | null {
```

Comment (jerader) [reply labware-library/src/labware-creator/components/sections/StackingOffsets.tsx]: oops i am always the worst at this, thank you!

File: labware-library/src/labware-creator/components/sections/StackingOffsets.tsx [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+    errors,
+    touched,
+    setFieldValue,
+  } = useFormikContext<LabwareFields>()
+
+  if (isEveryFieldHidden(fieldList, values)) {
+    return null
+  }
+
+  const label = `Stacking Offset (Optional)`
```
Comment (koji) [inline labware-library/src/labware-creator/components/sections/StackingOffsets.tsx]: nit

```suggestion
  const label = 'Stacking Offset (Optional)'
```

File: labware-library/src/labware-creator/components/sections/StackingOffsets.tsx [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+  }
+  if (isFlatBottom && isCircular) {
+    modifiedAdapterDefinitions = adapterDefinitions.filter(
+      definition =>
+        definition.parameters.loadName === 'opentrons_96_flat_bottom_adapter'
+    )
+  }
+  let modifiedModuleModels = MODULE_MODELS_WITH_NO_ADAPTERS
+  if (
+    (labwareHeight != null && parseInt(labwareHeight) > 16.06) ||
```
Comment (koji) [inline labware-library/src/labware-creator/components/sections/StackingOffsets.tsx]: nit

maybe define a const for `16.06`

File: labware-library/src/labware-creator/components/sections/StackingOffsets.tsx [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+          <div className={styles.instructions_column}>
+            <p>
+              Stacking offset is required for labware to be placed on modules
+              and adapters. Measure from the bottom of the adapter to the
+              highest part of the labware using a pair of calipers.
+            </p>
+          </div>
+          {modifiedAdapterDefinitions.length === 0 ? null : (
+            <Flex gridGap={SPACING.spacing4} flexDirection={DIRECTION_COLUMN}>
+              <StyledText as="h3" fontWeight={600}>
```
Comment (koji) [inline labware-library/src/labware-creator/components/sections/StackingOffsets.tsx]:
```suggestion
              <StyledText as="h3" fontWeight={TYPOGRAPHY.fontWeightSemiBold}>
```

File: labware-library/src/labware-creator/components/sections/StackingOffsets.tsx [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+                  </Flex>
+                )
+              })}
+            </Flex>
+          )}
+          {isTiprack ? null : (
+            <>
+              <Box
+                borderBottom={`1px solid ${COLORS.grey30}`}
+                marginY="1.5rem"
```
Comment (koji) [inline labware-library/src/labware-creator/components/sections/StackingOffsets.tsx]:
```suggestion
                marginY={SPACING.spacing24}
```

File: labware-library/src/labware-creator/components/sections/StackingOffsets.tsx [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+              />
+              <Flex
+                gridGap={SPACING.spacing4}
+                flexDirection={DIRECTION_COLUMN}
+                marginTop={SPACING.spacing4}
+              >
+                <StyledText as="h3" fontWeight={600}>
```
Comment (koji) [inline labware-library/src/labware-creator/components/sections/StackingOffsets.tsx]:
```suggestion
                <StyledText as="h3" fontWeight={TYPOGRAPHY.fontWeightSemiBold}>
```

File: labware-library/src/labware-creator/fieldsToLabware.ts [hunk 10]
```diff
@@ -103,6 +104,31 @@ export function fieldsToLabware(
...
+        adapterDefinitions != null
+          ? Object.values(adapterDefinitions).find(
+              definition => definition.parameters.loadName === loadName
+            )?.dimensions.zDimension ?? 0
+          : 0
+      return (stackingOffsetWithLabware[loadName] = {
+        x: 0,
+        y: 0,
+        //  ensure that z is a number!
+        z: fields.labwareZDimension + adapterHeight - parseFloat(z as any),
```
Comment (koji) [inline labware-library/src/labware-creator/fieldsToLabware.ts:119]: need `parseFloat`? 
Seems that z is always `number`.

File: labware-library/src/labware-creator/fieldsToLabware.ts [hunk 11]
```diff
@@ -103,6 +104,31 @@ export function fieldsToLabware(
...
+      })
+    })
+    const stackingOffsetWithModule: Record<string, LabwareOffset> = {}
+    Object.entries(compatibleModules).forEach(([moduleModel, z]) => {
+      return (stackingOffsetWithModule[moduleModel] = {
+        x: 0,
+        y: 0,
+        //  ensure that z is a number!
+        z: fields.labwareZDimension - parseFloat(z as any),
```
Comment (koji) [inline labware-library/src/labware-creator/fieldsToLabware.ts:128]: same question

File: labware-library/src/labware-creator/fieldsToLabware.ts [hunk 12]
```diff
@@ -160,6 +186,14 @@ export function fieldsToLabware(
             : { displayCategory: groupMetadataDisplayCategory }),
         },
       },
+      stackingOffsetWithLabware:
+        Object.keys(stackingOffsetWithLabware).length > 0
+          ? stackingOffsetWithLabware
+          : undefined,
```
Comment (koji) [inline labware-library/src/labware-creator/fieldsToLabware.ts:192]: Do you think it's possible to use `null` instead of `undefined`?

**Comment by koji** [comment]
I could not uncheck these check boxes

Magnetic Block GEN1 shows up instead of Thermocycler GEN2 (wellPlate)

Screenshot 2024-06-07 at 12 57 07 PM

File: labware-library/src/labware-creator/fieldsToLabware.ts (hunk 10 above)
Comment (jerader) [reply labware-library/src/labware-creator/fieldsToLabware.ts:119]: apparently the `TextField` turns the input field into a string even though i specify it as a number? so this adds extra protection but technically you are right, it should always be a number.

**Review by koji — APPROVED** [review]
lgtm!

**Comment by jerader** [comment]
this is the followup ticket to add the image: https://opentrons.atlassian.net/browse/AUTH-499

**Pull request merged**